File: PA1/List.java
//-----------------------------------------------------------------------------
// Georgio Klironomos
// gklirono
// Programming Assignment 1
// List ADT for CMPS 101
//-----------------------------------------------------------------------------
//WORKS CITED
//http://javadevnotes.com/java-string-array-examples
//https://alvinalexander.com/java/java-string-array-reference-java-5-for-loop-syntax
//https://www.geeksforgeeks.org/doubly-linked-list/
//https://classes.soe.ucsc.edu/cmps101/Fall18/Examples/Java/Queue/Queue.java

class List{

    private class Node{
      // Fields
      int data;
      Node next;
      Node prev;
      
      // Constructor
      Node(int data) { this.data = data; next = null; prev = null;}
      
      // toString():  overrides Object's toString() method
      public String toString() {
         return String.valueOf(data);
      }
      
   }
   
   //Fields
   private Node front;
   private Node back;
   private Node cursor;
   private int length;
   private int index;

    //constructor
    List() {
        front = back = cursor = null;
        length = 0;
        index = -1;
    }    
    //access functions
    int length() {
        return length;
    }
    int index() {
        return index;
    }

    int front() {
        if(length > 0) {
            return front.data; 
		}
		else {return -1;}
    }
    int back() {
        if(length > 0) {
            return back.data; 
		}
		else {return -1;}
    }
    int get() {
		if(index >= 0) {
                return cursor.data; 
		}
		else {return -1;}
    }
    boolean equals(List L) {
        //true if and only this List and passed-in L are the same integer sequence.
        //while loop using both cursor and comparing each element, maybe
        if(L.length != length){
            return false;
        }
        else{
            List.Node current1 = front;
            List.Node current2 = L.front;
            while (current1 != null){
                if(current1.data != current2.data){
                    return false;
                }    
                current1 = current1.next;
                current2 = current2.next;    
            }    
            return true;
        }
    }
    //manipulation procedures
    void clear() {
        while(length > 0){
            deleteFront();
        }
    }
    void moveFront() {
        if(length > 0) {
            cursor = front;
            index = 0;
        }
    }
    void moveBack() {
        if(length > 0) {
            cursor = back;
            index = (length-1);
        }
    }
    void movePrev() {
        if(cursor != null){
            cursor = cursor.prev;
            index--;
        }
    }
    void moveNext()  {
        if(cursor != null){
            cursor = cursor.next;
            if(index == (length-1)) {index = -1;}
            else{index++;}
        }
    }
    void prepend(int data) {
        List.Node new_node = new List.Node(data);
		if(length()>0){
			new_node.next = front;
			front.prev = new_node;
			front = new_node;
		}
		if(length() == 0){
	    	front = new_node;
	    	back = new_node;
		}
        length++;
        if(cursor != null){
            index++;//NEW!
        }
    }
    void append(int data) {
        List.Node new_node = new List.Node(data);		
        if(length() > 0){	
            back.next = new_node;				
            new_node.prev = back;				
            back = new_node;				
        }   
        if(length() == 0){
            front = new_node;
            back = new_node;
        }
        length++;
    }
    void insertBefore(int data) {
        if(length > 0 && index >= 0) {
            List.Node new_node = new List.Node(data);
            new_node.next = cursor;
            if(cursor.prev != null) {
                new_node.prev = cursor.prev;
                cursor.prev.next = new_node;
            }
            cursor.prev = new_node;
            if(cursor == front){
                front = new_node;
            }
            length++;
            index++;
        }
    }    
    void insertAfter(int data) {
        if(length > 0 && index >= 0) {
            List.Node new_node = new List.Node(data);
            new_node.prev = cursor;
            if(cursor.next != null) {
                new_node.next = cursor.next;
                cursor.next.prev = new_node;
            }
            cursor.next = new_node;
            if(cursor == back){
                back = new_node;
            }
            length++;
        }
    }
    void deleteFront() {
        if (length <= 0 || front == null){
            return;
        }
        if (front.next!=null){
            List.Node frontNext = front.next;
            frontNext.prev = null;
        }
        if(cursor == front){
            cursor = null;
            index = -1;
        }
        front = front.next;
        length--;
    }
    void deleteBack() {
        if (length <= 0 || back == null){
            return;
        }
        if (back.prev!=null){
            List.Node backPrev = back.prev;
            backPrev.next = null;
        }
        if(cursor == back){
            cursor = null;
            index = -1;
        }
        back = back.prev;
        length--;
    }    
    void delete() {
        if (length <= 0 || length <0 || cursor == null){
            return;
        }
        if(cursor.prev!=null){
            List.Node cursorPrev = cursor.prev;
            cursorPrev.next = cursor.next;
            if(back == cursor){
                back = cursorPrev;
            }
        }
        if(cursor.next!=null){
            List.Node cursorNext = cursor.next;
            cursorNext.prev = cursor.prev;
             if(front == cursor){
                front = cursorNext;
            }
        }
        cursor = null;
        index = -1;
        length--;
    }
    //other methods
    public String toString(){
        String listString = "";
        List.Node current = front;
        while(current != null){
            listString += Integer.toString(current.data) +" ";
            current = current.next;
        }
        String cursorStr = "null";
        if (cursor != null) {
            cursorStr = Integer.toString(cursor.data);
        }
        return listString;
    }
    List copy() {
        List newList = new List();
        List.Node current = front;
        while(current != null) {
            newList.append(current.data);
            current = current.next;
        }
        return newList;
    }
    List concat(List L) {
        List newList = new List();
        List.Node current = front;
        while(current != null) {
            newList.append(current.data);
            current = current.next;
        }
        current = L.front;
        while(current != null) {
            newList.append(current.data);
            current = current.next;
        }
        return newList;
    }
}

File: PA2/List.c
//-----------------------------------------------------------------------------
// List.c
// Implementation file for List ADT
//-----------------------------------------------------------------------------

/*WORKS CITED
https://classes.soe.ucsc.edu/cmps101/Fall18/Examples/C/Queue/Queue.c
https://stackoverflow.com/questions/10668504/segmentation-fault-c
https://en.cppreference.com/w/cpp/language/operator_incdec
https://stackoverflow.com/questions/2612447/pinpointing-conditional-jump-or-move-depends-on-uninitialized-values-valgrin
*/

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// private NodeObj type
typedef struct NodeObj{
   int data;
   struct NodeObj* next;
   struct NodeObj* prev;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private ListObj type
typedef struct ListObj{
   Node front;
   Node back;
   Node cursor;
   int length;
   int index;
} ListObj;


// Constructors-Destructors ---------------------------------------------------

// newNode()
// Returns reference to new Node object. Initializes next, prev,  and data fields.
// Private.
Node newNode(int data, Node prev, Node next){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->prev = prev;
   N->next = next;
   return(N);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}

// newList()
// Returns reference to new empty List object.
List newList(void){
   List L = malloc(sizeof(ListObj));
   L->front = L->back = L->cursor = NULL;
   L->length = 0;
   L->index = -1;
   return(L);
}


// freeList()
// Frees all heap memory associated with List *pL, and sets *pL to NULL
void freeList(List* pL){
   if(pL!=NULL && *pL!=NULL) { 
       Node current1 = (*pL)->front;
      while(current1 != NULL) { 
        Node current2 = current1;
        current1 = current1->next;
        freeNode(&current2);
        current2 = NULL;
      }
      free(*pL);
      *pL = NULL;
   }
}


// Access functions -----------------------------------------------------------

// getLength()
// Returns the length of L.
int length(List L){
   if( L==NULL ){
      printf("List Error: calling length() on NULL List reference\n");
      exit(1);
   }
   return(L->length);
}

int index(List L) {
    return(L->index);
}

// getFront()
// Returns the value at the front of L.
// Pre: !isEmpty(L)
int front(List L){
   if( (L==NULL) || (L->length < 1) ){
    printf("List Error: calling front() on NULL List reference\n");
    exit(1);
   }
   return(L->front->data);
}

int back(List L){
   if( (L==NULL) || (L->length < 1) ){
    printf("List Error: calling back() on NULL List reference\n");
    exit(1);
   }
   return(L->back->data);
}

int get(List L) {
   if((L == NULL) || (L->index < 0) || (L->length < 1)) {
      printf("List Error: calling get() on NULL List or NULL cursor\n");
      exit(1);
   }
   return(L->cursor->data);
}

int equals(List A, List B) {
    int same = 0;
   Node N = NULL;
   Node M = NULL;
   if( A==NULL || B==NULL ){
      printf("List Error: calling equals() on NULL List reference\n");
      exit(1);
   }
   same = ( A->length == B->length );
   N = A->front;
   M = B->front;
   while( same && N!=NULL){
      same = (N->data==M->data);
      N = N->next;
      M = M->next;
   }
   return same;
}

void clear(List L) {
    Node current1 = L->front; 
    while(current1 != NULL) {
      Node current2 = current1;
      current1 = current1->next;
      freeNode(&current2);
    }
    L->front = L->back = L->cursor = NULL;
    L->length = 0;
    L->index = -1;
}

void moveFront(List L) {
  if(L->length > 0) {
    L->cursor = L->front;
    L->index = 0;
  }
}
void moveBack(List L) {
  if(L->length > 0) {
    L->cursor = L->back;
    L->index = L->length-1;
  }
}
void movePrev(List L) {
  if(L->cursor != NULL){
            L->cursor = L->cursor->prev;
            L->index--;
        }
}
void moveNext(List L) {
  if(L->cursor != NULL){
            L->cursor = L->cursor->next;
            if(L->index == (L->length-1)) {L->index = -1;}
            else{L->index++;}
        }
}
void prepend(List L, int data) {
  Node new_node = newNode(data, NULL, L->front);
		if(L->length > 0){
			L->front->prev = new_node;
		}
		if(L->length == 0){
	    	L->back = new_node;
		}
    L->front = new_node;  
    L->length++;
    if(L->cursor != NULL){
      L->index++;//NEW!
    }
}

void append(List L, int data) {
  Node new_node = newNode(data, NULL, NULL);		
        if(L->length > 0){	
            L->back->next = new_node;
            new_node->prev = L->back;										
        }   
        if(L->length == 0){
            L->front = new_node;
        }
        L->back = new_node;
        L->length++;
}

void insertBefore(List L, int data) {
  if(L->length > 0 && L->index >= 0) {
            Node new_node = newNode(data, L->cursor->prev, L->cursor);
            if(L->cursor->prev != NULL) {
                L->cursor->prev->next = new_node;
            }
            L->cursor->prev = new_node;
            if(L->cursor == L->front){
                L->front = new_node;
            }
            L->length++;
            L->index++;
        }
}

void insertAfter(List L, int data) {
  if(L->length > 0 && L->index >= 0) {
            Node new_node = newNode(data, L->cursor, L->cursor->next);
            if(L->cursor->next != NULL) {
                L->cursor->next->prev = new_node;
            }
            L->cursor->next = new_node;
            if(L->cursor == L->back){
                L->back = new_node;
            }
            L->length++;
        }
}

void deleteFront(List L) {
  if (L->length <= 0 || L->front == NULL){
            return;
        }
        if (L->front->next!=NULL){
            L->front->next->prev = NULL;
        }
        if(L->cursor == L->front){
            L->cursor = NULL;
            L->index = -1;
        }
        else {
            L->index--;
        }
        free(L->front);
        L->front = L->front->next;
        L->length--;
}

void deleteBack(List L) {
  if (L->length <= 0 || L->back == NULL){
            return;
        }
        if (L->back->prev!=NULL){
            L->back->prev->next = NULL;
        }
        if(L->cursor == L->back){
            L->cursor = NULL;
            L->index = -1;
        }
        free(L->back);
        L->back = L->back->prev;
        L->length--;
}

void delete(List L) {
  if (L->length <= 0 || L->length <0 || L->cursor == NULL) {
            return;
        }
        if(L->cursor->prev!=NULL) {
            L->cursor->prev->next = L->cursor->next;
            if(L->back == L->cursor){
                L->back = L->cursor->prev;
            }
        }
        if(L->cursor->next!=NULL){
            L->cursor->next->prev = L->cursor->prev;
             if(L->front == L->cursor){
                L->front = L->cursor->next;
            }
        }
        free(L->cursor);
        L->cursor = NULL;
        L->index = -1;
        L->length--;
}

void printList(FILE* out, List L) {
  Node current = L->front;
  while(current != NULL){
    fprintf(out, "%d ",current->data); 
    current = current->next;   
  }
  free(current);  
}

List copyList(List L) {
    List newbie = newList();
    if (L->length > 0 && L->front != NULL) {
        Node current = L->front;
        while(current != NULL) {
            append(newbie, current->data);
            current = current->next;
        }
    }
    
    return newbie;
}